src/workerd/api/message-channel.h

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+#pragma once
+
+#include <workerd/api/basics.h>
+#include <workerd/jsg/jsg.h>
+
+#include <kj/array.h>
+
+namespace workerd::api {
+
+// Implements MessagePort web-spec
+// Ref: https://html.spec.whatwg.org/multipage/web-messaging.html#message-ports
+class MessagePort: public EventTarget {
+ public:
+  MessagePort() = default;
+  ~MessagePort() noexcept(false);
+  KJ_DISALLOW_COPY(MessagePort);
+
+  static jsg::Ref<MessagePort> constructor();
+
+  struct StructuredSerializeOptions {
+    kj::Array<jsg::Object> transfer{};
+
+    JSG_STRUCT(transfer);
+  };
+
+  void postMessage(jsg::Lock& js,
+      jsg::Value message,
+      kj::OneOf<kj::Maybe<StructuredSerializeOptions>, kj::Array<jsg::Value>> options);
+  void start(jsg::Lock& js);
+  void stop(jsg::Lock& js);
+  void close(jsg::Lock& js);
+
+  JSG_RESOURCE_TYPE(MessagePort) {
+    JSG_NESTED_TYPE(EventTarget);
+    JSG_METHOD(postMessage);
+    JSG_METHOD(start);
+    JSG_METHOD(stop);
+    JSG_METHOD(close);
+  }
+
+  // Ref: https://html.spec.whatwg.org/multipage/web-messaging.html#disentangle
+  void disentangle(jsg::Lock& js);
+
+  // Ref: https://html.spec.whatwg.org/multipage/web-messaging.html#entangle
+  void entangle(jsg::Lock& js, jsg::Ref<MessagePort> port);
+
+ private:
+  bool detached = false;
+
+  kj::Maybe<jsg::HashableV8Ref<v8::Object>> onmessage;
+  kj::Maybe<jsg::HashableV8Ref<v8::Object>> onmessageerror;
+
+  // Each MessagePort object can be entangled with another (a symmetric relationship)
+  kj::Maybe<jsg::Ref<MessagePort>> entangledWith{};
```

Comment:
You'll need the `visitForGc` method implemented in this class also visiting each of these fields (`onmessage`, `onmessageerror`, and `entagledWith`)